File: train/ReplayBuffer.py
from torch.utils.data import Dataset


class ReplayBuffer(Dataset):
    def __init__(self, max_size=None):
        self.training_data = []
        self.max_size = max_size

    def add_samples(self, new_samples):
        self.training_data.extend(new_samples)
        if self.max_size is not None and len(self.training_data) > self.max_size:
            self.training_data = self.training_data[-self.max_size:]

    def size(self):
        return len(self.training_data)

    def __getitem__(self, index):
        return self.training_data[index]

    def __len__(self):
        return len(self.training_data)

File: ReplayBuffer.cpp
#include "ReplayBuffer.h"

ReplayBuffer::ReplayBuffer(size_t max_size) : max_size(max_size) {}

void ReplayBuffer::add_samples(const std::vector<Sample>& new_samples) {
    training_data.insert(training_data.end(), new_samples.begin(), new_samples.end());
    if (max_size != 0 && training_data.size() > max_size) {
        training_data.erase(training_data.begin(), training_data.begin() + (training_data.size() - max_size));
    }
}

torch::optional<size_t> ReplayBuffer::size() const {
    return training_data.size();
}

Sample ReplayBuffer::get(size_t index) {
    return training_data[index];
}